src/clr-core/button/button.element.ts

```diff
@@ -84,82 +79,59 @@ export class CwcButton extends CwcBaseButton {
    * - `error`: shows the content of the button (in the context of application, this state is usually entered from a LOADING state. the application should show appropriate error message)
    */
   @property({ type: String })
-  set loadingState(state: ClrLoadingState) {
-    const oldState = this._loadingState;
-    this.updateButtonState(state);
-    this.requestUpdate('loadingState', oldState);
-  }
+  loadingState = ClrLoadingState.DEFAULT;
+
+  firstUpdated(props: Map<string, any>) {
+    super.firstUpdated(props);
 
-  connectedCallback() {
-    super.connectedCallback();
-    this.tabIndex = 0;
+    if (this.loadingState !== ClrLoadingState.DEFAULT) {
+      this.updateLoadingState();
+    }
   }
 
-  get loadingState() {
-    return this._loadingState;
+  update(props: Map<string, any>) {
+    if (this.privateHost && props.has('loadingState')) {
+      this.updateLoadingState(props.get('loadingState'));
+    }
+    super.update(props);
   }
 
   render() {
-    switch (this.loadingState) {
-      case ClrLoadingState.LOADING:
-        return html`
-          <span class="spinner spinner-inline"></span>
-          ${this.hiddenButtonTemplate}
-        `;
-      case ClrLoadingState.SUCCESS:
-        return html`
-          <span class="spinner spinner-inline spinner-check"></span>
-          ${this.hiddenButtonTemplate}
-        `;
-      default:
-        return html`
-          <slot></slot>
-          ${this.hiddenButtonTemplate}
-        `;
-    }
+    return html`
+    <div class="private-host">
+      ${this.loadingState === ClrLoadingState.LOADING ? iconSpinner : ''}
+      ${this.loadingState === ClrLoadingState.SUCCESS ? iconSpinnerCheck : ''}
+      ${this.loadingState === ClrLoadingState.DEFAULT ? html`<slot></slot>` : ''}
+      ${this.hiddenButtonTemplate}
+    </div>
+    `;
   }
 
   static get styles() {
     return [baseStyles, styles];
   }
 
-  private updateButtonState(state: ClrLoadingState) {
-    if (state === this.loadingState) {
-      return;
-    }
-
-    this._loadingState = state;
-
-    if (state === ClrLoadingState.ERROR) {
-      // in case of error, go back to default case; app should display relevant error message
-      this.loadingState = ClrLoadingState.DEFAULT;
+  private updateLoadingState(previousState?: ClrLoadingState) {
+    if (this.loadingState === ClrLoadingState.LOADING) {
+      this.privateHost.style.width = getElementWidth(this.privateHost);
+      this.disabled = true;
+    } else if (this.loadingState === ClrLoadingState.SUCCESS && previousState === ClrLoadingState.LOADING) {
+      setTimeout(() => {
```

Comment:
Is it necessary to check for `previousState === ClrLoadingState.LOADING` here? What happens when the button enters `SUCCESS` from `DEFAULT` state? Shouldn't the button still be disabled in this case?

Maybe a broader discussion would be - do we make certain assumptions about what the previous state is? How do we want to handle cases like what I mentioned? I think the tricky part is that there is an aspect of the button that doesn't depend on previous state (disabled/enabled) and one that does (setting the width of the button to prevent button resizing).

Reply:
➕ 1

Why do we care about the previous loading state? The only reason I can imagine us wanting to use it is in a situation where we don't want to do anything if the new state and the previous state are the same and, iirc, Lit handles that for us.

IMO, ln 118 should only check for `this.loadingState === ClrLoadingState.SUCCESS`. Managing the previous state or the cycle of states is out of scope for the type of state we want to build into these components.

Reply:
Ah sorry that should not be there, it was leftover from some debugging I believe. I'll get that removed